src/mumble/pluginAPI/Plugin.h

```diff
@@ -0,0 +1,242 @@
+	/// 	freshly connected to the server) or the channel isn't available because of any other reason.
+	/// @param newChannelID The ID of the channel the user has entered. If the ID is negative, the new channel could not be retrieved. This means
+	/// 	that the ID is invalid.
+	PLUGIN_EXPORT void onChannelEntered(MumbleConnection_t connection, MumbleUserID_t userID, MumbleChannelID_t previousChannelID,
+			MumbleChannelID_t newChannelID);
+
+	/// Called whenever a user leaves a channel.
+	/// This function will also be called when the user disconnects from the server.
+	///
+	/// @param connection The ID of the server-connection this event is connected to
+	/// @param userID The ID of the user that left the channel
+	/// @param channelID The ID of the channel the user left. If the ID is negative, the channel could not be retrieved. This means that the ID is
+	/// 	invalid.
+	PLUGIN_EXPORT void onChannelExited(MumbleConnection_t connection, MumbleUserID_t userID, MumbleChannelID_t channelID);
+
+	/// Called when any user changes his/her talking state.
+	///
+	/// @param connection The ID of the server-connection this event is connected to
+	/// @param userID The ID of the user whose talking state has been changed
+	/// @param talkingState The new TalkingState the user has switched to.
+	PLUGIN_EXPORT void onUserTalkingStateChanged(MumbleConnection_t connection, MumbleUserID_t userID, TalkingState_t talkingState);
+
+	/// Called whenever there is audio input.
+	///
+	/// @param inputPCM A pointer to a short-array holding the pulse-code-modulation (PCM) representing the audio input.
+	/// @param inputSize The size of the input-array
+	/// @param isSpeech A boolean flag indicating whether Mumble considers the input as part of speech (instead of background noise)
+	/// @returns Whether this callback has modified the audio input-array
+	PLUGIN_EXPORT bool onAudioInput(short *inputPCM, uint32_t inputSize, bool isSpeech);
+
+	/// Called whenever there is audio output.
+	///
+	/// @param outputPCM A pointer to a short-array holding the pulse-code-modulation (PCM) representing the audio output
+	/// @param sampleCount How many samples (array-entries) there are for each channel
+	/// @param channelCount How many channels there are in the audio output
+	/// @returns Whether this callback has modified the audio output-array
+	PLUGIN_EXPORT bool onAudioOutput_short(short *outputPCM, uint32_t sampleCount, uint32_t channelCount);
+
+	/// Called whenever there is audio output.
+	///
+	/// @param outputPCM A pointer to a float-array holding the pulse-code-modulation (PCM) representing the audio output
+	/// @param sampleCount How many samples (array-entries) there are for each channel
+	/// @param channelCount How many channels there are in the audio output
+	/// @returns Whether this callback has modified the audio output-array
+	PLUGIN_EXPORT bool onAudioOutput_float(float *outputPCM, uint32_t sampleCount, uint32_t channelCount);
```

Comment:
Same as above.

Reply:
Additionally, do we have access to the audio when it is first pulled off the incoming codec stream, and after the audio is expanded to the required number of output channels?

Reply:
> do we have access to the audio when it is first pulled off the incoming codec stream

That wasn't intended (yet). If required I could add this kind of functionality though (I guess)

> after the audio is expanded to the required number of output channels

Didn't even know such an expansion is happening - so dunno yet xD